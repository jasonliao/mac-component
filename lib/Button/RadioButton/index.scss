.radio-button {
  display: flex;
  align-items: center;
  cursor: default;

  .radio-button__box {
    display: inline-block;
    background: #fff;
    border-width: 1px;
    border-style: solid;
    border-color: #c8c8c8 #d7d7d7 #d2d2d2;
    border-radius: 50%; 
    box-shadow: inset 0px 0.4px 0.1px 0px rgba(0, 0, 0, 0.2);
    box-sizing: border-box;
    margin-right: 10px;
    position: relative;
  }
}

.radio-button:active {
  .radio-button__box {
    background: #f0f0f0;
    border-color: #c3c3c3 #cacaca #ccc;
  }
}

.radio-button__checked {
  .radio-button__box {
    background: #3B99FC;
    border: none;

    &::after {
      content: '';
      position: absolute;
      background: #fff;
      border-radius: 50%;
      box-shadow: 0px 0.5px 1px 0px #217FE2;
    }
  }
}

.radio-button__checked:active {
  .radio-button__box {
    background: #0075DB;
    border: none;
  }
}

.radio-button__regular {
  font-size: 13px;
  .radio-button__box {
    width: 16px;
    height: 16px;

    &::after {
      width: 6px;
      height: 6px;
      top: 5px;
      left: 5px;
    }
  }
}

.radio-button.radio-button__disabled,
.radio-button.radio-button_disabled:active {
  .radio-button__box {
    background: #f6f6f6;
    border-color: #d7d7d7 #dedede #dbdbdb;
  }
  .radio-button__label {
    color: #bebebe;
  }
}

.radio-button__small {
  font-size: 11px;
  .radio-button__box {
    width: 14px;
    height: 14px;

    &::after {
      width: 4px;
      height: 4px;
      top: 5px;
      left: 5px;
    }
  }
}

.radio-button__mini {
  font-size: 9px;
  .radio-button__box {
    width: 11px;
    height: 11px;

    &::after {
      width: 3px;
      height: 3px;
      top: 4px;
      left: 4px;
    }
  }
}