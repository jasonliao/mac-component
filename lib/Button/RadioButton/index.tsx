import * as React from 'react'
import classNames from 'classnames'

import CheckAndRadio, { CheckAndRadioProps } from '../BaseComponents/CheckAndRadio'

import './index.scss'

export interface RadioButtonProps extends CheckAndRadioProps {}

export default class RadioButton extends CheckAndRadio<RadioButtonProps> {
  render() {
    const { size, disabled, className, children, style } = this.props
    const hasChecked = this.hasChecked()

    const cls = classNames(
      className,
      'radio-button',
      `radio-button__${size}`,
      {
        'radio-button__checked': hasChecked,
        'radio-button__disabled': disabled
      }
    )

    return (
      <div
        className={cls}
        style={style}
        onClick={this.handleClick}
      >
        <i className="radio-button__box"></i>
        <span className="radio-button__label">{ children }</span>
      </div>
    )
  } 
}
