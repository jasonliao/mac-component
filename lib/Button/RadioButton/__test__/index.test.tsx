import * as React from 'react'
import { shallow } from 'enzyme'
import * as renderer from 'react-test-renderer'

import RadioButton, { RadioButtonProps } from '../index'

describe('RadioButton Component', () => {
  const label = 'Radio'
  it('[UI] render label', () => {
    const wrapper = shallow(<RadioButton>{ label }</RadioButton>)

    const text = wrapper.find('.radio-button__label').text()
    expect(text).toBe(label)
  })

  it('[UI] render with mini size', () => {
    const props: RadioButtonProps = { size: 'mini' }
    const wrapper = shallow(<RadioButton { ...props } />)

    const hasClass = wrapper.find('.radio-button').hasClass(`radio-button__${props.size}`)
    expect(hasClass).toBe(true)
  })

  it('[UI] render disabled radio', () => {
    const props: RadioButtonProps = { disabled: true }
    const wrapper = shallow(<RadioButton { ...props } />)

    const hasClass = wrapper.find('.radio-button').hasClass('radio-button__disabled')
    expect(hasClass).toBe(true)
  })

  it('[BEHAVIOR] radio clicked(uncontrolled)', () => {
    const props: RadioButtonProps = { onChange: jest.fn() }
    const wrapper = shallow(<RadioButton { ...props } />)
    const stateChecked = wrapper.state('checked')
  
    wrapper.simulate('click')
    expect(props.onChange).toHaveBeenCalledTimes(1)
    expect(wrapper.state('checked')).toBe(!stateChecked)
  })

  it('[BEHAVIOR] disabled radio clicked(uncontrolled)', () => {
    const props: RadioButtonProps = {
      disabled: true,
      onChange: jest.fn()
    }
    const wrapper = shallow(<RadioButton { ...props } />)
    const stateChecked = wrapper.state('checked')

    wrapper.simulate('click')
    expect(props.onChange).toHaveBeenCalledTimes(0)
    expect(wrapper.state('checked')).toBe(stateChecked)
  })

  it('[BEHAVIOR] radio clicked(controlled)', () => {
    const props: RadioButtonProps = {
      onChange: jest.fn(),
      checked: false
    }
    const wrapper = shallow(<RadioButton { ...props } />)
    const stateChecked = wrapper.state('checked')

    wrapper.simulate('click')
    expect(props.onChange).toHaveBeenCalledTimes(1)
    expect(wrapper.state('checked')).toBe(stateChecked)
  })

  it('[BEHAVIOR] disabled radio clicked(controlled)', () => {
    const props: RadioButtonProps = {
      onChange: jest.fn(),
      checked: false,
      disabled: true
    }
    const wrapper = shallow(<RadioButton { ...props } />)
    const stateChecked = wrapper.state('checked')

    wrapper.simulate('click')
    expect(props.onChange).toHaveBeenCalledTimes(0)
    expect(wrapper.state('checked')).toBe(stateChecked)
  })

  it('[SNAPSHOT] match the snapshot', () => {
    const props: RadioButtonProps = {
      size: 'small',
      checked: true,
      style: { width: '130px' },
      className: 'custom_classname'
    }
    const tree = renderer.create(<RadioButton { ...props }>{ label }</RadioButton>)
    expect(tree).toMatchSnapshot()

  })
})