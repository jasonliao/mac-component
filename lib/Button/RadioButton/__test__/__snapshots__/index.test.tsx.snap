// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButton Component [SNAPSHOT] match the snapshot 1`] = `
<div
  className="custom_classname radio-button radio-button__small radio-button__checked"
  onClick={[Function]}
  style={
    Object {
      "width": "130px",
    }
  }
>
  <i
    className="radio-button__box"
  />
  <span
    className="radio-button__label"
  >
    Radio
  </span>
</div>
`;
